agent/php_environment.c

```diff
@@ -511,6 +511,110 @@ static void nr_php_get_environment_variables(TSRMLS_D) {
                    __func__, NR_PHP_PROCESS_GLOBALS(env_labels));
 }
 
+#define MAX_LINE_COUNT (1000)  // Upper bound for number of lines to read
+/*
+ * Purpose:
+ *    Extract the 64-byte hexadecimal Docker cgroup ID from
+ *     /proc/self/mountinfo
+ */
+char* nr_php_parse_v2_docker_id(const char* cgroup_fname) {
+  char* line_ptr = NULL;
+  char* retval = NULL;
+  bool found = false;
+  int line_count = 0;
+  FILE* fd = NULL;
+  size_t len = 0;
+  nr_regex_t* line_regex = NULL;
+  nr_regex_substrings_t* ss = NULL;
+
+  if (NULL == cgroup_fname) {
+    return NULL;
+  }
+
+  // check if file exists
+  if (SUCCESS != access(cgroup_fname, F_OK)) {
+    nrl_verbosedebug(NRL_AGENT, "%s: File not found: %s", __func__,
+                     cgroup_fname);
+    return NULL;
+  }
+
+  // open file
+  fd = fopen(cgroup_fname, "r");
+  if (NULL == fd) {
+    nrl_warning(NRL_AGENT, "%s: Failed to open %s", __func__, cgroup_fname);
+    return NULL;
+  }
+
+  // compile regex to extract target string from file line
+  line_regex = nr_regex_create("/docker/containers/([a-fA-F0-9]{64})/", 0, 0);
+
+  if (NULL == line_regex) {
+    nrl_error(NRL_AGENT, "%s: Error: line regex creation failed", __func__);
+    fclose(fd);
+    return NULL;
+  }
+
+  // clang-format off
+  /* 
+   * Example /proc/self/mountinfo file structure:
+   *   ...
+   *   795 787 254:1 /docker/containers/ec807d5258c06c355c07e2acb700f9029d820afe5836d6a7e19764773dc790f5/resolv.conf /etc/resolv.conf rw,relatime - ext4 /dev/vda1 rw
+   *   796 787 254:1 /docker/containers/ec807d5258c06c355c07e2acb700f9029d820afe5836d6a7e19764773dc790f5/hostname /etc/hostname rw,relatime - ext4 /dev/vda1 rw
+   *   797 787 254:1 /docker/containers/ec807d5258c06c355c07e2acb700f9029d820afe5836d6a7e19764773dc790f5/hosts /etc/hosts rw,relatime - ext4 /dev/vda1 rw
+   *   ...
+   */
+
+  /*
+   * File parsing logic:
+   *  1. scan file line-by-line
+   *  2. regex search each line for '/docker/containers/' string followed by a 64 byte hex value
+   *  3. extract the 64 byte field following "/docker/containers/"
+   *  4. Verify the extracted ID is both:
+   *    a. 64 bytes long (not including null terminator)
+   *    b. comprised of only hexadecimal characters
+   *  5. Assign the extracted & verified ID to the retval
+   *    a. Example ID: ec807d5258c06c355c07e2acb700f9029d820afe5836d6a7e19764773dc790f5
+   *  6. Set found = true and exit the loops
+   */
+  // clang-format on
+
+  while (FAILURE != getline(&line_ptr, &len, fd) && !found
+         && line_count++ < MAX_LINE_COUNT) {
+    ss = nr_regex_match_capture(line_regex, line_ptr, nr_strlen(line_ptr));
+    if (NULL == ss) {
+      continue;
+    }
+    retval = nr_regex_substrings_get(ss, 1);
+    found = true;
+  }
+
+  nr_regex_destroy(&line_regex);
+  nr_regex_substrings_destroy(&ss);
```

Comment:
This looks SOOO much cleaner, great job!

Technically we can destroy it here because once we've found it we exist (since we set found to true) so regex_match_capture would only have been called once, but it might be better just to clean it after the retval is set since every regex_match_capture call creates an object that needs to be destroyed.

Reply:
Like this? be0cd8a4094d5e40ac5dc093f117a4af2dd2b239